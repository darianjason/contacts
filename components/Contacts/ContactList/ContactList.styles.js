import {StyleSheet} from 'react-native';
import {Colors} from '../../../constants';

const styles = StyleSheet.create({
  listContainer: {
    justifyContent: 'center',
    paddingHorizontal: 16,
    paddingVertical: 8,
  },
  altContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  emptyMessage: {
    color: Colors.primary,
    fontSize: 16,
  },
});

export default styles;
