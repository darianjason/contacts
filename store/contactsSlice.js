import {createSlice} from '@reduxjs/toolkit';

const contactsSlice = createSlice({
  name: 'contacts',
  initialState: {
    contacts: [],
  },
  reducers: {
    setContacts: (state, action) => {
      state.contacts = action.payload;
    },
    insertContact: (state, action) => {
      state.contacts.push(action.payload);
    },
    updateContact: (state, action) => {
      const {id, data} = action.payload;
      const updatedContactIndex = state.contacts.findIndex(
        contact => contact.id === id,
      );
      state.contacts[updatedContactIndex] = data;
    },
    deleteContact: (state, action) => {
      const {id} = action.payload;
      const index = state.contacts.findIndex(contact => contact.id === id);
      state.contacts.splice(index, 1);
    },
  },
});

const {actions, reducer} = contactsSlice;
export const {setContacts, insertContact, updateContact, deleteContact} =
  actions;
export default reducer;

const BASE_URL = 'https://simple-contact-crud.herokuapp.com';

export const fetchContacts = () => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact`);
    const contacts = await response.json();
    dispatch(setContacts(contacts.data));
  } catch (error) {
    throw error;
  }
};

export const addContact = contact => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(contact),
    });

    const responseJson = await response.json();
    console.log(responseJson);

    dispatch(insertContact(contact));
  } catch (error) {
    throw error;
  }
};

export const editContact = (id, data) => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    });

    const responseJson = await response.json();
    console.log(responseJson);

    dispatch(updateContact(id, data));
  } catch (error) {
    throw error;
  }
};

export const removeContact = id => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact/${id}`, {
      method: 'DELETE',
    });

    const responseJson = await response.json();
    console.log(responseJson);

    dispatch(deleteContact(id));
  } catch (error) {
    throw error;
  }
};
