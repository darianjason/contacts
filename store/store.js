// @flow

import {configureStore} from '@reduxjs/toolkit';

import contactsReducer from './contactsSlice';

type Contact = {
  id?: string,
  firstName: string,
  lastName: string,
  age: number | string,
  photo: string,
};

type Store = {
  contacts: Array<Contact>,
};

export const store: Store = configureStore({
  reducer: {
    contacts: contactsReducer,
  },
});
