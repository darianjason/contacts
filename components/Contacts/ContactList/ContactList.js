import React, {useState, useEffect} from 'react';
import {
  FlatList,
  View,
  ScrollView,
  ActivityIndicator,
  RefreshControl,
} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import {useDispatch, useSelector} from 'react-redux';

import {ContactListItem} from '..';
import {fetchContacts} from '../../../store/contactsSlice';
import {DefaultText} from '../../ui';
import {Colors} from '../../../constants';
import styles from './ContactList.styles';

const loadContacts = async (dispatch, setIsLoading) => {
  setIsLoading(true);
  try {
    await dispatch(fetchContacts());
  } catch (error) {
    console.error(error);
  }
  setIsLoading(false);
};

const selectHandler = (navigation, contact) => {
  navigation.navigate('ContactDetails', {contact});
};

const sortByName = (a, b) => {
  const nameA = a.firstName.toUpperCase() + a.lastName.toUpperCase();
  const nameB = b.firstName.toUpperCase() + b.lastName.toUpperCase();

  if (nameA < nameB) {
    return -1;
  }

  if (nameA > nameB) {
    return 1;
  }

  return 0;
};

const renderListItem = (navigation, itemData) => {
  const {firstName, lastName, age, photo} = itemData.item;

  return (
    <ContactListItem
      firstName={firstName}
      lastName={lastName}
      age={age}
      photo={photo}
      onSelect={() => selectHandler(navigation, itemData.item)}
    />
  );
};

const LoadingIndicator = () => (
  <View style={styles.altContainer}>
    <ActivityIndicator size="large" color={Colors.primary} />
  </View>
);

const EmptyMessage = ({dispatch, isLoading, setIsLoading, setError}) => (
  <ScrollView
    contentContainerStyle={styles.altContainer}
    refreshControl={
      <RefreshControl
        refreshing={isLoading}
        onRefresh={() => loadContacts(dispatch, setIsLoading, setError)}
      />
    }>
    <DefaultText style={styles.emptyMessage}>No contacts</DefaultText>
  </ScrollView>
);

const ContactList = () => {
  const contacts = useSelector(state => state.contacts.contacts);
  const dispatch = useDispatch();
  const navigation = useNavigation();

  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    loadContacts(dispatch, setIsLoading);
  }, [dispatch, setIsLoading]);

  if (isLoading) {
    return <LoadingIndicator />;
  }

  if (contacts.length === 0) {
    return (
      <EmptyMessage
        dispatch={dispatch}
        isLoading={isLoading}
        setIsLoading={setIsLoading}
      />
    );
  }

  return (
    <FlatList
      data={[...contacts].sort(sortByName)}
      renderItem={itemData => renderListItem(navigation, itemData)}
      onRefresh={() => loadContacts(dispatch, setIsLoading)}
      refreshing={isLoading}
      contentContainerStyle={styles.listContainer}
    />
  );
};

export default ContactList;
