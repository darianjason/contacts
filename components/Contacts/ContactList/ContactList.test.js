import React from 'react';
import {render} from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';

import {store} from '../../../store/store';
import ContactList, {sortByName} from './ContactList';

describe('Contact List component', () => {
  test('renders empty message if no contacts', async () => {
    const contacts = [];
    const response = {data: contacts};

    global.fetch = jest.fn();
    fetch.mockResolvedValueOnce({
      status: 200,
      ok: true,
      json: async () => response,
    });

    const {findByText} = render(
      <Provider store={store}>
        <ContactList />
      </Provider>,
    );

    const emptyMessage = await findByText('No contacts');

    expect(emptyMessage).toBeTruthy();
  });

  test('renders contacts if request succeeds', async () => {
    const contacts = [
      {
        id: '93ad6070-c92b-11e8-b02f-cbfa15db428b',
        firstName: 'Bilbo',
        lastName: 'Baggins',
        age: 111,
        photo:
          'http://vignette1.wikia.nocookie.net/lotr/images/6/68/Bilbo_baggins.jpg/revision/latest?cb=20130202022550',
      },
      {
        id: 'b3abd640-c92b-11e8-b02f-cbfa15db428b',
        firstName: 'Luke',
        lastName: 'Skywalker',
        age: 20,
        photo: 'N/A',
      },
    ];
    const response = {data: contacts};

    global.fetch = jest.fn();
    fetch.mockResolvedValueOnce({
      status: 200,
      ok: true,
      json: async () => response,
    });

    const {findAllByA11yLabel} = render(
      <Provider store={store}>
        <ContactList />
      </Provider>,
    );

    const contactList = await findAllByA11yLabel('Contact list item');

    expect(contactList).toHaveLength(2);
  });

  test('renders error message if request fails', async () => {
    const contacts = [
      {
        id: '93ad6070-c92b-11e8-b02f-cbfa15db428b',
        firstName: 'Bilbo',
        lastName: 'Baggins',
        age: 111,
        photo:
          'http://vignette1.wikia.nocookie.net/lotr/images/6/68/Bilbo_baggins.jpg/revision/latest?cb=20130202022550',
      },
      {
        id: 'b3abd640-c92b-11e8-b02f-cbfa15db428b',
        firstName: 'Luke',
        lastName: 'Skywalker',
        age: 20,
        photo: 'N/A',
      },
    ];
    const response = {message: 'Error message', data: contacts};

    global.fetch = jest.fn();
    fetch.mockResolvedValueOnce({
      status: 400,
      ok: false,
      json: async () => response,
    });

    const {findByText} = render(
      <Provider store={store}>
        <ContactList />
      </Provider>,
    );

    const errorMessage = await findByText(
      'Failed to fetch contacts: Error message (400)',
    );

    expect(errorMessage).toBeTruthy();
  });

  test('sorts alphabetically by full name', () => {
    const names = [
      {
        firstName: 'Zoe',
        lastName: 'Smith',
      },
      {
        firstName: 'A',
        lastName: 'Name',
      },
      {
        firstName: 'John',
        lastName: 'Doe',
      },
      {
        firstName: 'Hello',
        lastName: 'World',
      },
    ];
    const sortedNames = names.sort(sortByName);

    expect(sortedNames).toEqual([
      {
        firstName: 'A',
        lastName: 'Name',
      },
      {
        firstName: 'Hello',
        lastName: 'World',
      },
      {
        firstName: 'John',
        lastName: 'Doe',
      },
      {
        firstName: 'Zoe',
        lastName: 'Smith',
      },
    ]);
  });

  test('renders correctly', () => {
    const contacts = [
      {
        id: '93ad6070-c92b-11e8-b02f-cbfa15db428b',
        firstName: 'Bilbo',
        lastName: 'Baggins',
        age: 111,
        photo:
          'http://vignette1.wikia.nocookie.net/lotr/images/6/68/Bilbo_baggins.jpg/revision/latest?cb=20130202022550',
      },
      {
        id: 'b3abd640-c92b-11e8-b02f-cbfa15db428b',
        firstName: 'Luke',
        lastName: 'Skywalker',
        age: 20,
        photo: 'N/A',
      },
    ];
    const response = {data: contacts};

    global.fetch = jest.fn();
    fetch.mockResolvedValueOnce({
      status: 200,
      ok: true,
      json: async () => response,
    });

    const tree = renderer
      .create(
        <Provider store={store}>
          <ContactList />
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
