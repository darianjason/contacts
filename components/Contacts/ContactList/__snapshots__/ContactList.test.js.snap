// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact List component renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "justifyContent": "center",
      "paddingHorizontal": 16,
      "paddingVertical": 8,
    }
  }
  data={
    Array [
      Object {
        "age": 111,
        "firstName": "Bilbo",
        "id": "93ad6070-c92b-11e8-b02f-cbfa15db428b",
        "lastName": "Baggins",
        "photo": "http://vignette1.wikia.nocookie.net/lotr/images/6/68/Bilbo_baggins.jpg/revision/latest?cb=20130202022550",
      },
      Object {
        "age": 20,
        "firstName": "Luke",
        "id": "b3abd640-c92b-11e8-b02f-cbfa15db428b",
        "lastName": "Skywalker",
        "photo": "N/A",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  refreshing={false}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityLabel="Contact list item"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderRadius": 8,
              "flexDirection": "row",
              "marginVertical": 8,
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#E2EEF3",
                "borderRadius": 8,
                "height": 111.16666666666667,
                "justifyContent": "center",
                "overflow": "hidden",
                "width": 111.16666666666667,
              }
            }
          >
            <Image
              accessibilityRole="image"
              source={
                Object {
                  "uri": "http://vignette1.wikia.nocookie.net/lotr/images/6/68/Bilbo_baggins.jpg/revision/latest?cb=20130202022550",
                }
              }
              style={
                Object {
                  "height": "100%",
                  "width": "100%",
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "padding": 16,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#060A0E",
                  "fontFamily": "Manrope-Regular",
                }
              }
            >
              Bilbo
               
              Baggins
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityLabel="Contact list item"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderRadius": 8,
              "flexDirection": "row",
              "marginVertical": 8,
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#E2EEF3",
                "borderRadius": 8,
                "height": 111.16666666666667,
                "justifyContent": "center",
                "overflow": "hidden",
                "width": 111.16666666666667,
              }
            }
          >
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#142B3E",
                  "fontFamily": "Manrope-Bold",
                  "fontSize": 16,
                  "letterSpacing": 2,
                }
              }
            >
              L
              S
            </Text>
          </View>
          <View
            style={
              Object {
                "padding": 16,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#060A0E",
                  "fontFamily": "Manrope-Regular",
                }
              }
            >
              Luke
               
              Skywalker
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
